### ISSUE TITLE
A couple feature proposals (willing to implement, but want feedback first)

### ISSUE THREAD

**Issue by rwblokzijl** [pyauth/requests-http-signature#17]
Hello!

I am working on a project involving requests-oauthlib and this library. And i have run into some bugs/issues. For these i have provided possible fixes that i am willing to make a PR for after having recieved feedback.

**1 An empty body leads to error:**
`  File "[omitted]/python3.7/site-packages/requests/structures.py", line 54, in __getitem__
    return self._store[key.lower()][1]
KeyError: 'digest'
`
**My current workaround:** Override `add_digest`, and build in check:
```
class MyHTTPSignatureAuth(HTTPSignatureAuth):
    def add_digest(self, request):
        if not request.body:
            # Since the message body (payload) is empty, the digest is defaulted to the hash256 of an empty body.
            request.headers["Digest"] = "SHA-256=47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU="
        else:
            return super(MyHTTPSignatureAuth, self).__call__(request)
```
**Suggested change:** Build check into the actual code

**2 Incorrect body encoding leads to error:**
`
File "[omitted]/python3.7/site-packages/requests_http_signature/__init__.py", line 73, in add_digest
    digest = self.hasher_constructor(request.body).digest()
TypeError: Unicode-objects must be encoded before hashing
`
**My current workaround:** Override `add_digest`, and encode to utf-8:
```
class MyHTTPSignatureAuth(HTTPSignatureAuth):
    def add_digest(self, request):
        if request.body:
            request.body=request.body.encode('utf-8')
        super(MyHTTPSignatureAuth, self).add_digest(request)
```
**Suggested change:** This workaround probably wouldn't translate into a fix directly as `utf-8` is not always the desired encoding (i think). Though i dont understand enough about the crypto and http standard to be sure of this. If multiple encodings should be supported, the option should be added somewhere, if not, we should always convert to `utf-8`
**Note:** [In python3 everything is unicode](https://docs.python.org/release/3.0.1/whatsnew/3.0.html#text-vs-data-instead-of-unicode-vs-8-bit) thus this error will happen by default, always.

**3 Doesn't play nice with token authorization**
This auth lib works when requesting auth tokens because is sets the [`Authorization` header](https://tools.ietf.org/html/draft-cavage-http-signatures-12#section-3.1.2). However when signing subsequent messages, the `Authorization` header should be the token, and the Signature should be specified in the [`Signature` header](https://tools.ietf.org/html/draft-cavage-http-signatures-12#section-4.1.1)
**My current workaround:** Override `__call__`, and build check for preexisting `Authorization` field:
```
class MyHTTPSignatureAuth(HTTPSignatureAuth):
    def __call__(self, request):
        if "Authorization" in request.headers:
            auth = request.headers["Authorization"]
            req = super(MyHTTPSignatureAuth, self).__call__(request)
            req.headers["Signature"]=req.headers["Authorization"].split(' ', 1)[1]
            req.headers["Authorization"]=auth
            return req
        else:
            return super(MyHTTPSignatureAuth, self).__call__(request)
```
**Suggested change:** Unclear. I don't know if adding the `Signature` field even is the intended purpose of the library. I don't know when the `Signature` header should be used over the `Authorization` header. For a fix i would probably lean towards providing an option in the constructor to set `Signature` instead of `Authorization`, or maybe go with the dirty code in the work around if it doen't cause any issues.

Any thoughts on these changes? Should i implement them in a PR?

P.s: here is the whole class i am using at the moment:
```
class MyHTTPSignatureAuth(HTTPSignatureAuth):
    def __call__(self, request):
        if "Authorization" in request.headers:
            auth = request.headers["Authorization"]
            req = super(MyHTTPSignatureAuth, self).__call__(request)
            req.headers["Signature"]=req.headers["Authorization"].split(' ', 1)[1]
            req.headers["Authorization"]=auth
            return req
        else:
            return super(MyHTTPSignatureAuth, self).__call__(request)


    def add_digest(self, request):
        if request.body:
            request.body=request.body.encode('utf-8')
            super(MyHTTPSignatureAuth, self).add_digest(request)
        else:
            super(MyHTTPSignatureAuth, self).add_digest(request)
            return
            # Since the message body (payload) is empty, the digest is defaulted to the hash256 of an empty body.
            request.headers["Digest"] = "SHA-256=47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU="
```

**Comment by Bloodyfool** [comment]
I now see **3** already has been solved (https://github.com/pyauth/requests-http-signature/issues/16)
However it leads me to now have to override 2 classes with the same `add_digest` function.

**Comment by kislyuk** [comment]
PRs are welcome.

Your issue 2 (encoding mismatch) does not sound like an issue to me. Just encode the body correctly with your desired encoding before passing it to requests.

I'll merge #14 asap.

**Comment by kislyuk** [comment]
To answer the question in your issue 3, you should use the `HTTPSignatureHeaderAuth` class if you wish to free up the `Authentication` header for other uses. With that said, you may want to revisit the design of your application since it usually does not make sense to use both HTTP Signature and token-based authorization. The RFC is hard to follow on this subject and should probably provide more guidance.

**Comment by Bloodyfool** [comment]
I agree that it is weird to require both, but some [API](https://developer.ing.com/openbanking/get-started/openbanking)s just require it.

**Comment by bhenneman** [comment]
Signatures can also be used for non-repudiation purposes, not only for authentication

**Comment by lynshi** [comment]
It doesn't seem like the changes in #16 have been published on PyPI; the package is still on v0.1.0 and the downloaded source still uses `signer` and `verifier`. Should we expect a release soon?

**Comment by kislyuk** [comment]
@lynshi the changes have been published.

**Comment by kislyuk** [comment]
After recent updates, all of the problems described in the original issue have been addressed.
